bytecomp/typeopt.ml

```diff
@@ -45,7 +45,21 @@ let maybe_pointer_type env ty =
 let maybe_pointer exp = maybe_pointer_type exp.exp_env exp.exp_type
 
 let array_element_kind env ty =
-  match scrape env ty with
+  let rty =
+    let sty = scrape env ty in
+    match sty with
+    | Tconstr (p, _, _) ->
+      begin
+        match Env.find_type p env with
+        | {type_unboxed = {unboxed = true; _}; _} ->
+          Misc.Stdlib.Option.value_default (fun x -> x.desc) ~default:sty
+            (Typedecl.get_unboxed_type_representation env ty)
```

Comment:
(Having `is_base_type` covered is useful for `Translcore.specialize_comparison`, for instance.)